admin/services/data-access/retro-input-assistant.data.service.js

```diff
@@ -0,0 +1,80 @@
+'use strict'
+
+const sqlService = require('./sql.service')
+const TYPES = require('./sql.service').TYPES
+const aaDataService = require('./access-arrangements.data.service')
+
+const service = {
+  /**
+   * @typedef {object} retroInputAssistantData
+   * @property {string} firstName
+   * @property {string} lastName
+   * @property {string} reason
+   * @property {number} checkId
+   * @property {number} pupilId
+   * @property {number} userId
+   */
+
+  /**
+   * @param {retroInputAssistantData} data
+   * @returns {Promise<any>}
+   */
+  create: async function create (data) {
+    const arrangementType = await aaDataService.sqlFindAccessArrangementsIdsWithCodes([aaDataService.CODES.RETRO_INPUT_ASSISTANT])
+    const retroInputAssistantTypeId = arrangementType[0].id
+
+    const insertData = {
+      pupil_id: data.pupilId,
+      recordedBy_user_id: data.userId,
+      accessArrangements_id: retroInputAssistantTypeId,
+      retroInputAssistantFirstName: data.firstName,
+      retroInputAssistantLastName: data.lastName,
+      retroInputAssistantReason: data.reason,
+      retroInputAssistant_check_id: data.checkId
+    }
+    return sqlService.create('pupilAccessArrangements', insertData)
+  },
+  /**
+   * @description looks up pupil id and current check id via url slug
+   * @param {string} pupilUrlSlug
+   * @returns {Promise<any>}
+   */
+  getPupilIdAndCurrentCheckIdByUrlSlug: function getPupilIdAndCurrentCheckIdByUrlSlug (pupilUrlSlug) {
+    const pupilIdSql = 'SELECT id, currentCheckId FROM mtc_admin.pupil WHERE urlSlug=@pupilUrlSlug'
+    return sqlService.readonlyQuery(pupilIdSql, [{
+      name: 'pupilUrlSlug',
+      type: TYPES.UniqueIdentifier,
+      value: pupilUrlSlug
+    }])
+  },
+  sqlFindEligiblePupilsBySchoolId: async function sqlFindEligiblePupilsBySchoolId (schoolId) {
+    const params = [
+      {
+        name: 'schoolId',
+        value: schoolId,
+        type: TYPES.Int
+      }
+    ]
+    const sql = `
+        SELECT
+            p.id,
+            p.foreName,
+            p.lastName,
+            p.middleNames,
+            p.dateOfBirth,
+            p.group_id,
+            p.urlSlug
+          FROM
+              [mtc_admin].[pupil] p
+              LEFT JOIN [mtc_admin].[pupilAccessArrangements] paa ON (paa.pupil_id = p.id)
+         WHERE p.school_id = @schoolId
+           AND p.attendanceId IS NULL
+           AND p.checkComplete = 1
+           AND paa.pupil_id IS NULL
+         ORDER BY lastName;
```

Comment:
Why do the ordering on the db when there is only a single instance of it?  Wouldn't it make more sense to order it on the web tired, which is designed for scaling?

Also, if the idea is to order it so as it pass it into the pupil identification service, this ordering is insufficient. From memory, it would need to be `surname`, `firstname`, `dateOfBirth`, `middleNames`.

Reply:
this is a mirror image of the function used in the main AA page autocomplete feature, except for the filtering (which only includes pupils with completed checks).

the pupil uuid is captured for the autocomplete, to be submitted with the form.

if both instances of this fall short then we should raise as a separate concern.